rp2040-hal/src/async_utils.rs

```diff
@@ -0,0 +1,144 @@
+//! Commonly used in async implementations.
+
+use core::{marker::PhantomData, task::Poll};
+
+pub(crate) mod sealed {
+    use core::{cell::Cell, task::Waker};
+    use critical_section::Mutex;
+
+    pub trait Wakeable: crate::typelevel::Sealed {
+        /// Returns the waker associated with driver instance.
+        fn waker() -> &'static IrqWaker;
+    }
+
+    /// This type wraps a `Waker` in a `Mutex<Cell<Option<_>>`.
+    ///
+    /// While `critical_section::Mutex` intregrates nicely with RefCell, RefCell addess a borrow
+    /// counter that is not necessary for this usecase.
+    ///
+    /// This type is kept sealed to prevent user from mistakenly messing with the waker such as
+    /// clearing it while the driver is parked.
+    pub struct IrqWaker {
+        waker: Mutex<Cell<Option<Waker>>>,
+    }
+    impl IrqWaker {
+        pub const fn new() -> Self {
+            Self {
+                waker: Mutex::new(Cell::new(None)),
+            }
+        }
+        pub fn wake(&self) {
+            critical_section::with(|cs| {
+                if let Some(waker) = self.waker.borrow(cs).take() {
+                    Waker::wake(waker);
+                }
+            });
+        }
+        pub fn register(&self, waker: &Waker) {
+            critical_section::with(|cs| {
+                let old = self.waker.borrow(cs).replace(Some(waker.clone()));
+                if old.map(|old| !old.will_wake(waker)).unwrap_or(false) {
+                    panic!("Only one task may be able to await on this driver.");
```

Comment:
I'm still nervous about this code. It's not that I can definitely say that there's something wrong. It's just that it's not obvious enough that it is correct.

And `will_wake` is allowed to return `false` whenever it wants. The docs say "This function works on a best-effort basis, and may return false even when the Wakers would awaken the same task." Couldn't that cause a panic?

Reply:
Another idea: Isn't it allowed to call `poll` with varying wakers?
Poll's doc seems to allow it and specifies what should happen: "Note that on multiple calls to poll, only the [Waker](https://doc.rust-lang.org/beta/core/task/struct.Waker.html) from the [Context](https://doc.rust-lang.org/beta/core/task/struct.Context.html) passed to the most recent call should be scheduled to receive a wakeup."
That could also trigger the `!old.will_wake(waker)` branch with the panic.

Reply:
> Isn't it allowed to call poll with varying wakers?
It could but all these wakers will be waking the same task otherwise more fundamental invariant have been broken, e.g. Only 1 future (and therefore task) may hold a mutable reference to an instance of i2c and be awaiting on that instance.

> "Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup."

Indeed, however "the Waker from the Context passed to the most recent call" is a grey concept. If an IRQ wakes a running future (eg during the poll_fn) the irq does not yet know about that most recent waker. If that had to be strongly upheld, [`Waker::clone_from`](https://doc.rust-lang.org/src/core/task/wake.rs.html#393-397) wouldn't make much sense to have (to point being to take the right take).

> Couldn't that cause a panic?
That could indeed 🤔 I might have forgotten about that. I had troubles with that while developing so that was helpful but if that's not reliable, we have not choice but to leave the user in the dark on that one.

Reply:
I removed the panic.

Reply:
Thank you, I never thought about such a usecase.
I think that’d fall indeed under the case where "will_wake" may or may not return true, so we’d have a spurious panic.
The panic may have triggered on the first or the second call, so quite unpredictable.